decisions/0004-sync-routes-from-k8s-to-cc.md

```diff
@@ -0,0 +1,113 @@
+
+### Kubernetes
+
+Kubernetes is also implemented via a series of "bulk sync loops" that are more commonly known as controllers, but there are some key differences in how these convergence loops work.
+1. k8s has a ["watch"](https://kubernetes.io/docs/reference/using-api/api-concepts/#efficient-detection-of-changes) API so clients clients can subscribe to incremental changes in desired and actual state
+1. [k8s.io/client-go/cache](https://godoc.org/k8s.io/client-go/tools/cache) provides a mechanism for clients to build and maintain local caches of kubernetes state
+1. [kubebuilder](https://book.kubebuilder.io/) Provides, amongst other things, an SDK for building reconciliation loops on top of watches and client side caching
+
+### Problems and Solutions
+
+To illustrate the necessity of these sync loops, it helps to think about how they improve resiliency in 2 cases of temporary network partition from the perspective of the CF API: 
+1. Problem: Updates to CCDB succeed ✅, but updates to the Route CRD fail ❌
+1. Problem: Updates to the Route CRD succeed ✅, but updates to CCDB fail ❌
+
+To handle these 2 failure modes, several approaches can be taken. 
+
+##### Mimic historical sync loops:
+1. Problem: CCDB ✅ & Route CRD ❌
+   1. Solution: Raise the error to the user, but allow the CCDB transaction to commit. A bulk sync loop will detect that the Route CRD is "missing" and create it later.
+1. Problem: Route CRD ✅ & CCDB ❌
+   1. Solution A: Raise the error to the user, and don't do the Route CRD write until CCDB changes are committed.
+   1. Solution B: A bulk sync loop will detect that the Route CRD is "extraneous" and delete it later.
+
+##### Forgo a bulk sync and attempt to bootstrap consistency with k8s off CCDB transactions:
+1. Problem: CCDB ✅ & Route CRD ❌
+   1. Solution: Raise the error to the user, abort the CCDB transaction.
+1. Problem: Route CRD ✅ & CCDB ❌
+   1. Solution: Roll back changes to the Route CRD... (but what if that still fails due to network partition?!)
+
+##### Sync from k8s to CCDB
+1. Problem: CCDB ✅ & Route CRD ❌
+   1. Solution: Raise the error to the user, abort the CCDB transaction.
+1. Problem: Route CRD ✅ & CCDB ❌
+   1. Solution: Raise the error to the user, but leave the Route CRD. The "bulk sync loop" will not consider the route "reconciled" until it is correct in CCDB.
+
+### Viable Long Term Implementations & Tradeoffs
+
+##### CCDB->k8s Sync in Ruby
+1. Mimics ProcessesSync for efficient access to CCDB
+1. Run it in an independent process from clock_global to avoid BOSH entanglement
+1. Unclear what scale limitations we might encounter if we regularly download
+   all k8s Routes.
+
+##### CCDB->k8s Sync in Go
+1. Probably a new, independent go program
+1. Can take advantage of go-client cache & reflector to avoid repeatedly
+   downloading all k8s state.
+   1. It's unclear, however, that kubebuilder could help
```

Comment:
I think it could, you'd just need to leverage the `Owns` aspect here:
https://godoc.org/sigs.k8s.io/controller-runtime/pkg/builder#Builder.Owns

Or maybe I'm missing something here

Reply:
I think what I'm getting at is that kubebuilder's reconcile-each-change pattern isn't useful for building bulk sync loops that download 2 global states (all the v3/routes and all the k8s Routes) and then reconcile them. It just reconciles one-at-a-time as things change, as far as I understand.

Reply:
I suppose it is a question of caching the v3/routes. At least as far as I understood, caching of k8s resources is a key point in kubebuilder's advice to handle changed resources one-by-one even for a bulk sync.

Reply:
I also think of the reconcile as an abstraction away from thinking about diffing the state. The reconcile can focus on handling that something has change and it has to wrangle the k8s state to look correct, while the source of the event handles diffing the Cloud Controller's state with the K8s state and sending reconciliation events for anything that is different.